blockdag/ghostdag.go

```diff
@@ -0,0 +1,132 @@
+package blockdag
+
+import "github.com/pkg/errors"
+
+func (dag *BlockDAG) selectedParentAnticone(node *blockNode) (*blockHeap, error) {
+	anticoneSet := newSet()
+	anticoneHeap := newUpHeap()
+	past := newSet()
+	var queue []*blockNode
+	for _, parent := range node.parents {
+		if parent == node.selectedParent {
+			continue
+		}
+		anticoneSet.add(parent)
+		queue = append(queue, parent)
+	}
+	for len(queue) > 0 {
+		var current *blockNode
+		current, queue = queue[0], queue[1:]
+		for _, parent := range current.parents {
+			if anticoneSet.contains(parent) || past.contains(parent) {
+				continue
+			}
+			isAncestorOfSelectedParent, err := dag.isAncestorOf(parent, node.selectedParent)
+			if err != nil {
+				return nil, err
+			}
+			if isAncestorOfSelectedParent {
+				past.add(parent)
+				continue
+			}
+			anticoneSet.add(parent)
+			anticoneHeap.Push(parent)
+			queue = append(queue, parent)
+		}
+	}
+	return &anticoneHeap, nil
+}
+
+// blueAnticoneSize returns the blue anticone size of 'block' from the worldview of 'context'.
+// Expects 'block' to be ∈ blue-set(context)
+func (dag *BlockDAG) blueAnticoneSize(block, context *blockNode) (uint32, error) {
+	for node := context; node != nil; node = node.selectedParent {
+		if blueAnticoneSize, ok := node.bluesAnticoneSizes[*block.hash]; ok {
+			return blueAnticoneSize, nil
+		}
+	}
+	return 0, errors.Errorf("block %s is not in blue-set of %s", block.hash, context.hash)
+}
+
+func (dag *BlockDAG) ghostdag(newNode *blockNode) (*blockHeap, error) {
+	newNode.selectedParent = newNode.parents.bluest()
+	newNode.bluesAnticoneSizes[*newNode.hash] = 0
+	newNode.blues = append(newNode.blues, newNode.selectedParent)
+	selectedParentAnticone, err := dag.selectedParentAnticone(newNode)
+	if err != nil {
+		return nil, err
+	}
+
+	for selectedParentAnticone.Len() > 0 {
+		blueCandidate := selectedParentAnticone.pop()
```

Comment:
If you are poping stuff out of selectedParentAnticone here, wouldn't it be empty by the time you want to return it?